src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -76,19 +76,30 @@ public static OrderedSet<Symbol> Analyze(CSharpCompilation compilation, MethodSy
                 foreach (var kvp in lazyDisallowedCaptures)
                 {
                     var variable = kvp.Key;
-                    var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
 
-                    if (variable is SynthesizedLocal local && local.SynthesizedKind == SynthesizedLocalKind.Spill)
+                    if (variable is LocalSymbol local)
                     {
-                        Debug.Assert(local.TypeWithAnnotations.IsRestrictedType());
-                        diagnostics.Add(ErrorCode.ERR_ByRefTypeAndAwait, local.GetFirstLocation(), local.TypeWithAnnotations);
+                        if (local.SynthesizedKind is SynthesizedLocalKind.Spill)
+                        {
+                            reportLocalAcrossAwaitError(diagnostics, local, local.GetFirstLocation());
```

Comment:
>reportLocalAcrossAwaitError(diagnostics, local, local.GetFirstLocation()); [](http://example.com/codeflow?start=28&length=74)

Could you give an example of a test that goes through this code path?

Reply:
For example `SpanStackSafetyTests.AsyncSpilling` would report the diagnostic at a different location (the synthesized local "declaration" instead of the problematic "await"):

```diff
++                 Diagnostic(ErrorCode.ERR_ByRefTypeAndAwait, "default(Span<int>)").WithArguments("System.Span<int>").WithLocation(17, 19)
--                 Diagnostic(ErrorCode.ERR_ByRefTypeAndAwait, "await I1()").WithArguments("System.Span<int>").WithLocation(17, 39)
```

Reply:
Makes sense, I will change this. I was just trying to preserve the existing behavior here, but I see it will be better to change this.